File: python/roman-numerals/roman_numerals.py
'''
This is not coded in python style,
but it's very readable and compact
'''
def numeral(number):
    roman = ''
    while number >= 1000: roman +=  'M'; number -= 1000
    if    number >= 900:  roman += 'CM'; number -= 900
    if    number >= 500:  roman +=  'D'; number -= 500
    if    number >= 400:  roman += 'CD'; number -= 400
    while number >= 100:  roman +=  'C'; number -= 100
    if    number >= 90:   roman += 'XC'; number -= 90
    if    number >= 50:   roman +=  'L'; number -= 50
    if    number >= 40:   roman += 'XL'; number -= 40
    while number >= 10:   roman +=  'X'; number -= 10
    if    number >= 9:    roman += 'IX'; number -= 9
    if    number >= 5:    roman +=  'V'; number -= 5
    if    number >= 4:    roman += 'IV'; number -= 4
    while number >= 1:    roman +=  'I'; number -= 1
    return roman

File: typescript/roman-numerals/roman-numerals.ts
// Map a roman numeral to the decimal value.
// Maps remember insertion order.
const R2D = new Map<string,number>()

R2D.set('M', 1000); R2D.set('CM', 900); R2D.set('D', 500); R2D.set('CD', 400)
R2D.set('C',  100); R2D.set('XC',  90); R2D.set('L',  50); R2D.set('XL',  40)
R2D.set('X',   10); R2D.set('IX',   9); R2D.set('V',   5); R2D.set('IV',   4)
R2D.set('I',    1)

export function toRoman(decimal: number): string {
  let roman = ''
  for (const [r, d] of R2D) {
    while (decimal >= d) {
      roman   += r
      decimal -= d
    }
  }
  return roman
}
